lwr_controllers/src/gravity_compensation.cpp

```diff
@@ -0,0 +1,83 @@
+
+namespace lwr_controllers 
+{
+    GravityCompensation::GravityCompensation() {}
+    GravityCompensation::~GravityCompensation() {}
+    
+    bool GravityCompensation::init(hardware_interface::PositionJointInterface *robot, ros::NodeHandle &n)
+    {
+        KinematicChainControllerBase<hardware_interface::PositionJointInterface>::init(robot, n);
+        
+        // find stiffness (dummy) joints; this is necessary until proper position/stiffness/damping interface exists
+        for(std::vector<KDL::Segment>::const_iterator it = kdl_chain_.segments.begin(); it != kdl_chain_.segments.end(); ++it)
+        {
+            joint_stiffness_handles_.push_back(robot->getHandle(it->getJoint().getName()+"_stiffness"));
+        }
+        
+        ROS_DEBUG("found %lu stiffness handles", joint_stiffness_handles_.size());
+        
+        sub_command_ = nh_.subscribe("command", 1, &GravityCompensation::command, this);
+        sub_stiffness_ = nh_.subscribe("set_stiffness", 1, &GravityCompensation::set_stiffness, this);
+        
+        previous_stiffness_.resize(joint_stiffness_handles_.size());
+        stiffness_ = DEFAULT_STIFFNESS;
+        enabled_ = true;
+        
+        return true;		
+    }
+    
+    void GravityCompensation::starting(const ros::Time& time)
+    {
+        for(size_t i=0; i<joint_handles_.size(); i++) 
+        {
+            previous_stiffness_[i] = joint_stiffness_handles_[i].getPosition();
+        }
+    }
+    
+    void GravityCompensation::update(const ros::Time& time, const ros::Duration& period)
+    {
+        if(enabled_)
+        {
+            // update the commanded position to the actual, so that the robot doesn't 
+            // go back at full speed to the last commanded position when the stiffness 
+            // is raised again
+            for(size_t i=0; i<joint_handles_.size(); i++) 
+            {
+                // clamp the measured position to the soft limits, in order to avoid instability
+                float cmd_position = joint_handles_[i].getPosition();
+                if (cmd_position < joint_limits_.min(i)) {
+                    ROS_INFO("clamping joint %i from %f to %f", i, cmd_position, joint_limits_.min(i));
+                    cmd_position = joint_limits_.min(i);
+                }
+                if (cmd_position > joint_limits_.max(i)) {
+                    ROS_INFO("clamping joint %i from %f to %f", i, cmd_position, joint_limits_.max(i));
+                    cmd_position = joint_limits_.max(i);
+                }
+                joint_handles_[i].setCommand(cmd_position);
+                joint_stiffness_handles_[i].setCommand(stiffness_);
+            }
+        } else {
+            for(size_t i=0; i<joint_handles_.size(); i++) 
+            {
+                joint_stiffness_handles_[i].setCommand(previous_stiffness_[i]);
+            }
+        }        
+    }
+    
+    void GravityCompensation::command(const std_msgs::Bool::ConstPtr &msg)
```

Comment:
dynamic_reconfigure would definitely be better than a FollowJointTrajectory interface, but still you would need joint handles so this would not solve the problem, right?